admin-dev/themes/new-theme/js/pages/product/components/combination-list/bulk-delete-handler.ts

```diff
@@ -100,49 +98,65 @@ export default class BulkDeleteHandler {
     });
   }
 
-  /**
-   * @todo: the bulk delete action should be displayed with a progress modal once it is ready.
-   */
   private async bulkDelete(combinationIds: number[]): Promise<void> {
-    const progressModal = this.showProgressModal();
-    const progressModalElement = document.getElementById(CombinationMap.bulkProgressModalId);
+    const $bulkDeleteBtn = $(CombinationMap.bulkDeleteBtn);
+    const bulkChunkSize = Number($bulkDeleteBtn.data('bulkChunkSize'));
+    const abortController = new AbortController();
+
+    const progressModal = new ProgressModal({
+      id: CombinationMap.bulkProgressModalId,
+      abortCallback: () => {
+        stopProcess = true;
+        abortController.abort();
+      },
+      closeCallback: () => this.eventEmitter.emit(CombinationEvents.bulkDeleteFinished),
+      progressionTitle: $bulkDeleteBtn.data('progressTitle'),
+      progressionMessage: $bulkDeleteBtn.data('progressMessage'),
+      closeLabel: $bulkDeleteBtn.data('closeLabel'),
+      abortProcessingLabel: $bulkDeleteBtn.data('stopProcessing'),
+      errorsMessage: $bulkDeleteBtn.data('errorsMessage'),
+      backToProcessingLabel: $bulkDeleteBtn.data('backToProcessing'),
+      downloadErrorLogLabel: $bulkDeleteBtn.data('downloadErrorLog'),
+      viewErrorLogLabel: $bulkDeleteBtn.data('viewErrorLog'),
+      viewErrorTitle: $bulkDeleteBtn.data('viewErrorTitle'),
+      total: combinationIds.length,
+    });
+    progressModal.show();
+    let stopProcess = false;
+    let doneCount = 0;
+    while (combinationIds.length) {
+      const chunkIds: number[] = combinationIds.splice(0, bulkChunkSize);
+
+      if (stopProcess) {
+        break;
+      }
 
-    let progress = 1;
+      let data: Record<string, any>;
 
-    //@todo: is there better loop option to avoid disabling eslint?
-    // eslint-disable-next-line no-restricted-syntax
-    for (const combinationId of combinationIds) {
       try {
         // eslint-disable-next-line no-await-in-loop
-        await this.combinationsService.deleteCombination(combinationId);
-      } catch (error) {
-        console.error(error);
+        data = await this.combinationsService.bulkDeleteCombinations(this.productId, chunkIds);
```

Comment:
Do we really need to block the call stack execution? I think you told me about this some days ago but maybe it could be worth it to handle every request in an async way and populate the progress bar/errors without blocking it

Reply:
Same answer as the last time, the blocking here is on purpose because we don't want 5 requests in parallel to avoid too much burden on the server So the idea is to have only one request at a time However each request handles a chunk of IDs so this single request can actually perform action on 5/10/whatever elements at a time

At least that's my opinion to avoid too many requests, but it would also be possible to handle several ones in parallel That's debatable but even if we do this I think it can be done later in a refacto Because it should still include some limitations about the number of requests that should be executed in parallel, so we'd need to stack each request to keep track on them and be able to abort them if needed

Reply:
If we don't wait for response, then progress will just continue even if there are errors (but we want to stop/show the errors for every request), so idk how else could we achieve it without waiting for every request to finish (i think that's the whole idea about the progress bar - to show live progress)